test/accel/UserActionIntegration.test.cc

```diff
@@ -83,6 +94,191 @@ TEST_F(LarSphere, run)
+{
+    auto result = LarSphereIntegrationMixin::make_physics_input();
+
+    // Set default optical physics
+    auto& optical = result.optical;
+    optical = {};
+
+    // Disable generation of Cherenkov and scintillation photons in Geant4
+    optical.cherenkov.stack_photons = false;
+    optical.scintillation.stack_photons = false;
+
+    // Disable WLS which isn't yet working (reemission) in Celeritas
+    using WLSO = WavelengthShiftingOptions;
+    optical.wavelength_shifting = WLSO::deactivated();
+    optical.wavelength_shifting2 = WLSO::deactivated();
+
+    return result;
+}
+
+//---------------------------------------------------------------------------//
+/*!
+ * Enable optical tracking with distribution offloading.
+ */
+auto LarSphereOpticalOffload::make_setup_options() -> SetupOptions
+{
+    auto result = LarSphereIntegrationMixin::make_setup_options();
+
+    result.optical_capacity = [] {
+        inp::OpticalStateCapacity cap;
+        cap.tracks = 32768;
+        cap.generators = cap.tracks * 8;
+        cap.primaries = cap.tracks * 16;
+        return cap;
+    }();
+
+    // Enable optical distribution offloading
+    result.optical_generator = inp::OpticalOffloadGenerator{};
+
+    // Don't offload any particles
+    result.offload_particles = SetupOptions::VecG4PD{};
+
+    return result;
+}
+
+//---------------------------------------------------------------------------//
+/*!
+ * Stepping action for pushing optical distributions to Celeritas.
+ */
+void LSOOSteppingAction::UserSteppingAction(G4Step const* step)
+{
+    CELER_EXPECT(step);
+
+    constexpr double clhep_time{1 / units::nanosecond};
+
+    auto& local = detail::IntegrationSingleton::local_optical_offload();
+    if (!local)
+    {
+        // Offloading is disabled
+        return;
+    }
+
+    if (step->GetStepLength() == 0)
+    {
+        // Skip "no-process"-defined steps
+        return;
+    }
+
+    auto* pm = step->GetTrack()->GetDefinition()->GetProcessManager();
+    CELER_ASSERT(pm);
+
+    // Determine how many Cherenkov and scintillation photons to generate
+    size_type num_cherenkov{0};
+    size_type num_scintillation{0};
+    if (auto const* p = dynamic_cast<G4Cerenkov const*>(pm->GetProcess("Cerenk"
+                                                                       "ov")))
+    {
+        num_cherenkov = p->GetNumPhotons();
+    }
+    if (auto const* p = dynamic_cast<G4Scintillation const*>(
+            pm->GetProcess("Scintillation")))
+    {
+        num_scintillation = p->GetNumPhotons();
+    }
+
+    if (num_cherenkov == 0 && num_scintillation == 0)
+    {
+        return;
+    }
+
+    auto* pre_step = step->GetPreStepPoint();
+    auto* post_step = step->GetPostStepPoint();
+    CELER_ASSERT(pre_step && post_step);
+
+    // Create distribution and push to Celeritas
+    // TODO: Get optical material ID
+    // TODO: Does the post-step speed account for only continuous energy
+    // loss or continuous+discrete?
```

Comment:
I think continuous plus discrete! @Rashika-Gupta @stognini this ties into our conversation today. We could probably use a new issue for how to accurately describe the scintillation/cherenkov due to along-step energy loss versus end-of-step energy deposition due to tracking cuts or (for example) the hard-coded 100 eV Klein-Nishina electron production threshold.